src/runtime/opengl.cpp

```diff
@@ -1941,23 +1959,34 @@ WEAK int halide_opengl_initialize_kernels(void *user_context, void **state_ptr,
         } else {
             len = next_kernel - this_kernel;
         }
-        
-        // Construct a new ModuleState and add it to the global list
-        module = (ModuleState *)malloc(sizeof(ModuleState));
-        module->kernel = NULL;
-        module->next = state_list;
-        state_list = module;
-        *state = module;
-        
-        KernelInfo *kernel = module->kernel;
-        if (!kernel) {
-            kernel = create_kernel(user_context, this_kernel, len);
-            if (!kernel) {
-                error(user_context) << "Invalid kernel: " << this_kernel;
-                return -1;
+        if (module->num_kernels == module->max_kernels) {
+            // allocate a new array
+            KernelInfo** new_kernels = (KernelInfo**)malloc(
+                sizeof(KernelInfo*) * (module->max_kernels + max_kernel_increment));
+
+            if (!module->kernel) {
+                error(user_context) << "Unable to allocate memory for more kernels";
             }
-            module->kernel = kernel;
+
+            // copy old kernel ptrs into new array
+            for (int i=0; i<module->num_kernels; i++) {
+                new_kernels[i] = module->kernel[i];
+            }
+            // dealloc old kernel array & replace in module
+            free(module->kernel);
+            module->kernel = new_kernels;
+            module->max_kernels += max_kernel_increment;
         }
+
+        KernelInfo *kernel = module->kernel[module->num_kernels];
+        kernel = create_kernel(user_context, this_kernel, len);
```

Comment:
Can combine the previous two lines... there's no point in initializing kernel twice.